### DISCUSSION TITLE
Strongswan VPN scalability

### DISCUSSION THREAD

**Discussion by harisasank** [strongswan/strongswan#2035] [Q&A] [answered]
I am trying to run strongswan with 12K VPNs.  The IKE negotiation seems to take a long time to come up (around 8 seconds / VPN on an Icelake system) when I run using load-tester at the otherend.  I am trying to see if we can do any better and the following are my initial observations
- I have set the IKE half open timeout to 240 seconds - otherwise I see a lot of invalid_spi messages. 
- During negotiation, charon-systemd, runs at around 3600% CPU ( I still have more cores available ) - and I am using around 128 threads. Using 512 threads did n't make much difference either. 
- As per perf report, most of the time, seems to spent inside processing IKE_SA_INIT finding a matching peer and doing an insertion sort on the linked list. I am trying to see if we can do anything here to improve this.  
+   98.37%     0.00%  charon-systemd  libcharon.so.0.0.0               [.] get_ike_cfg
+   98.26%     0.13%  charon-systemd  libcharon.so.0.0.0               [.] get_matching_ike_cfgs
+   92.27%    16.88%  charon-systemd  libcharon.so.0.0.0               [.] insert_sorted_ike
+   75.34%    29.86%  charon-systemd  libstrongswan.so.0.0.0           [.] enumerator_enumerate_default
+   47.81%    17.19%  charon-systemd  libstrongswan.so.0.0.0           [.] enumerate_next
+   33.93%    32.59%  charon-systemd  libstrongswan.so.0.0.0           [.] do_enumerate
In absolute terms, get_ike_cfg function in backend manager takes a lit bit over 2 whole seconds sometimes. and this is per IKE_SA_INIT message.

If somehave has faced a similar issue like this or have some insights of what can be done, I will really appreciate it. 

Thanks.

**Comment by harisasank** [comment]
I have attached sample configuration of connections.conf and the load-tester.conf below. My hunch is that, memory allocation of enumerator objects might be one of the bottlenecks here.
get_ike_cfg tries to create a linked list of 12K nodes, and then sort them based on the get_ike_match callback.


[connections.json](https://github.com/strongswan/strongswan/files/13671561/connections.json)

**Comment by tobiasbrunner** [comment]
You have 12K configs loaded? Why? What's different in these that would require that? strongSwan is definitely not optimized to handle that many configs (it can handle lots of SAs, but not lots of configs).

**Comment by harisasank** [comment]
Thanks for you reply.  I have a unique use case, where each VPN client (which is the ike initiator) will to use a different traffic selector in CHILD SA, but unfortunately our client will not be able to give that subnet list in the phase2 exchange. So, instead, each client will have its own VPN in the server that it will connect to using a unique ID and will have (0.0.0.0 as TSi), and the traffic selector will be determined in the server side (strongswan side). 
I understand that supporting this use case may not be of much interest to the strongswan, but I would still like to know if something can be done about the scalability issues related to VPN Configurations.

**Reply by tobiasbrunner** [reply]
Could you please clarify what you mean. Are you referring to virtual IPs (TSi is usually narrowed to that)? Or are you referring to the TS on the server side (which would be TSr)? In which case each client could only communicate with a different specific subnet/IP? What would the purpose be of that?

**Reply by harisasank** [reply]
The VPN clients have a limitation that prevents them from specifying traffic selectors in the exchange, they all send wildcard ts. The actual ts are known VPN server, that performs ts narrowing. The only way for the server to select the right CHILD_SA (and hence ts) is to have a separate connection for each client, identified by its identity or IP.

**Reply by tobiasbrunner** [reply]
Sorry, I don't understand what you mean or why you think this is the only option for your scenario. You also haven't answered my questions.

**Reply by harisasank** [reply]
Sorry about the confusion. 
-  In which case each client could only communicate with a different specific subnet/IP?
   Yes, In Client CREATE_CHILD_SA (TSi - 0.0.0.0; TSr - 0.0.0.0).
   The Server narrows it down as (TSi - a.b.c.d/24; TSr - a.b.c.e/24)
-   And also, if you think there is a better way to do this, I will be be glad to know.

**Reply by tobiasbrunner** [reply]
So each of your 12k clients gets different `TSi` *and* `TSr`, or is only one of them different for each client? And what about virtual IPs? (Your example there uses `/24` not `/32`.)

**Reply by harisasank** [reply]
Clients are not assigned a virtual IP, and each client has a fixed subnet. The client implementation is not able to specify this subnet in the TSi. Each client's TSi is therefore stored in the server's IKE configuration. 
- So each of your 12k clients gets different TSi and TSr, or is only one of them different for each client? 
  On the server, the local TS is 0.0.0.0 and the remote TS is the client-specific subnet. So any traffic coming to the server from a third IP, that is destined to a client-specific subnet will take the client-specific tunnel.

**Reply by tobiasbrunner** [reply]
Just to get this right, the `tunnel1` config you posted above is completely unrelated to all this then? That had `pools` set to assign virtual IPs to the client and it only specified `local_ts` (i.e. TSr) and not `remote_ts` (TSi, which would default to the virtual IP). So in your "real" configs, you don't set `pools` and set `local_ts=0.0.0.0/0` and `remote_ts=<client-specific subnet>`?

**Reply by harisasank** [reply]
Yes, that is correct. I was just trying to understand the general slow functions in strongswan there.

**Comment by tobiasbrunner** [comment, marked as answer]
OK, given what we discussed in [the other thread](https://github.com/strongswan/strongswan/discussions/2035#discussioncomment-7860970), you have two options to avoid loading a config for each client (i.e. use a single config for all clients):

1. If you use X.509 certificates for authentication (the config you posted above uses PSKs, but since that isn't the "real" config, I don't know if you do, if you currently don't, maybe consider it), use [RFC 3779](https://datatracker.ietf.org/doc/html/rfc3779) IPAddrBlocks extensions in the client (and optionally CA) certificates. The [_addrblock_ plugin](https://docs.strongswan.org/docs/5.9/plugins/addrblock.html) will use these to narrow the client's traffic selectors (TSi) to the addresses/subnets encoded in the client certificate. By default, the plugin will also check the encoded extensions against CA certificates, but that can be disabled via `depth` option.
2. The second option is to implement [`listener_t::narrow()`](https://github.com/strongswan/strongswan/blob/f3578d3de839815ed0dad17b0a3c3317365bd1fe/src/libcharon/bus/listeners/listener.h#L258-L271) in a custom plugin. That allows you to dynamically narrow the traffic selectors (TSi and/or TSr) based on the client's identity (or IP address or whatever else you want to use) e.g. using an in-memory hashtable or using a database (completely depends on the custom implementation). Such plugins can be built out-of-tree  and then loaded into the daemon (see [this example](https://github.com/tobiasbrunner/strongswan-redirect-plugin)).

**Reply by harisasank** [reply]
Thanks a lot for the pointers, Tobias.  Appreciate your help.
